Processes in the WebProcessCache should suspend more promptly

cdumez opened the pull request:
#### 02d6615aea2e51b0f048e9500e9f24565231d36e
<pre>
Processes in the WebProcessCache should suspend more promptly
<a href="https://bugs.webkit.org/show_bug.cgi?id=256468">https://bugs.webkit.org/show_bug.cgi?id=256468</a>
rdar://108624630

Reviewed by Geoffrey Garen.

When processes entered the WebProcessCache, they would not suspend for ~8 minutes
due to the near-suspended assertion we take after releasing the last useful
process assertion.

To cause them to suspend more promptly, I am adding an implementation for
WebProcessProxy::platformSuspendProcess() &amp; platformResumeProcess() to drop
&amp; retake the near-suspended assertion. platformSuspendProcess() gets called
by the WebProcessCache 30 seconds after the process has entered the cache.

This patch effectively causes our WebProcesses to fully suspend 30 seconds after
entering the WebProcessCache, instead of 8 minutes. This tested as neutral on
PLT5 and Membuster.

* Source/WebKit/UIProcess/API/Cocoa/WKProcessPool.mm:
(-[WKProcessPool _requestWebProcessTermination:]):
(-[WKProcessPool _isWebProcessSuspended:]):
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPoolPrivate.h:
* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::setShouldTakeNearSuspendedAssertion):
* Source/WebKit/UIProcess/ProcessThrottler.h:
(WebKit::ProcessThrottler::isHoldingNearSuspendedAssertion const):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/mac/WebProcessProxyMac.mm:
(WebKit::WebProcessProxy::platformSuspendProcess):
(WebKit::WebProcessProxy::platformResumeProcess):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:

Canonical link: <a href="https://commits.webkit.org/263839@main">https://commits.webkit.org/263839@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36ac37c83db883c383c7cc11722f19d7abf47ed3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5831 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5999 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6188 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7391 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6217 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5832 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6222 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5966 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/8312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5937 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5986 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5284 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7448 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3463 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5259 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13224 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5330 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5338 "5 failures") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7559 "262 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5784 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4749 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5225 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/1389 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9346 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5586 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren reviewed (approved):
r=me

bnham commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -338,6 +338,18 @@ void ProcessThrottler::setAllowsActivities(bool allow)
+    if (shouldTakeNearSuspendedAssertion) {
+        if (!m_assertion && m_processIdentifier) {
+            PROCESSTHROTTLER_RELEASE_LOG_WITH_PID("setShouldTakeNearSuspendedAssertion: Taking near-suspended assertion", m_processIdentifier);
+            setThrottleState(ProcessThrottleState::Suspended);
```
In the case where we're resuming a process out of the process cache, is it always guaranteed that we execute this `platformResumeProcess => setShouldTakeNearSuspendedAssertion(true)` path before some other code calls `setThrottleState(ProcessThrottleState::Foreground)` on the process?

cdumez replied:
platformResumeProcess() is the first thing that will get called when we take a process out of the process cache, before we do anything with this process. So yes, we'll definitely call `setShouldTakeNearSuspendedAssertion(true)` before some other code calls `setThrottleState(ProcessThrottleState::Foreground)`. That said, I believe it would also be fine if we didn't. It is fine to go from no assertion to a foreground assertion. There is no requirement to go from suspended assertion to foreground assertion.

bnham reviewed (commented):
Thought: I have a bit of trouble following the ProcessThrottler state machine these days. Maybe all the functions that want to manipulate throttle state (such as `setShouldTakeSuspendedAssertion`) could call a function like `recalculateThrottleState` and all the logic that handles what to do with `m_assertion` could be in the body of `setThrottleState`? This is just a thought and doesn't have to block this PR from being merged.

cdumez committed:
Processes in the WebProcessCache should suspend more promptly
https://bugs.webkit.org/show_bug.cgi?id=256468
rdar://108624630

Reviewed by Geoffrey Garen.

When processes entered the WebProcessCache, they would not suspend for ~8 minutes
due to the near-suspended assertion we take after releasing the last useful
process assertion.

To cause them to suspend more promptly, I am adding an implementation for
WebProcessProxy::platformSuspendProcess() & platformResumeProcess() to drop
& retake the near-suspended assertion. platformSuspendProcess() gets called
by the WebProcessCache 30 seconds after the process has entered the cache.

This patch effectively causes our WebProcesses to fully suspend 30 seconds after
entering the WebProcessCache, instead of 8 minutes. This tested as neutral on
PLT5 and Membuster.

* Source/WebKit/UIProcess/API/Cocoa/WKProcessPool.mm:
(-[WKProcessPool _requestWebProcessTermination:]):
(-[WKProcessPool _isWebProcessSuspended:]):
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPoolPrivate.h:
* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::setShouldTakeNearSuspendedAssertion):
* Source/WebKit/UIProcess/ProcessThrottler.h:
(WebKit::ProcessThrottler::isHoldingNearSuspendedAssertion const):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/mac/WebProcessProxyMac.mm:
(WebKit::WebProcessProxy::platformSuspendProcess):
(WebKit::WebProcessProxy::platformResumeProcess):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:

Canonical link: https://commits.webkit.org/263839@main
